dg.go

```diff
@@ -1,315 +1,487 @@
+		}
+		for _, nodeID := range removeNodeIDs {
+			for targetNodeID := range connectionsToNode {
+				delete(connectionsToNode[targetNodeID], nodeID)
+			}
+		}
+	}
 }
 
 func (d *directedGraph[NodeType]) AddNode(id string, item NodeType) (Node[NodeType], error) {
-    d.lock.Lock()
-    defer d.lock.Unlock()
-    if _, ok := d.nodes[id]; ok {
-        return nil, ErrNodeAlreadyExists{
-            id,
-        }
-    }
-    d.nodes[id] = &node[NodeType]{
-        false,
-        id,
-        item,
-        d,
-    }
-    d.connectionsToNode[id] = map[string]struct{}{}
-    d.connectionsFromNode[id] = map[string]struct{}{}
-    return d.nodes[id], nil
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	if _, ok := d.nodes[id]; ok {
+		return nil, ErrNodeAlreadyExists{
+			id,
+		}
+	}
+	d.nodes[id] = &node[NodeType]{
+		deleted:                 false,
+		ready:                   false,
+		id:                      id,
+		item:                    item,
+		status:                  Waiting,
+		outstandingDependencies: make(map[string]DependencyType),
+		dg:                      d,
+	}
+	d.connectionsToNode[id] = map[string]struct{}{}
+	d.connectionsFromNode[id] = map[string]struct{}{}
+	return d.nodes[id], nil
 }
 
 func (d *directedGraph[NodeType]) GetNodeByID(id string) (Node[NodeType], error) {
-    d.lock.Lock()
-    defer d.lock.Unlock()
-
-    n, ok := d.nodes[id]
-    if !ok {
-        return nil, &ErrNodeNotFound{
-            id,
-        }
-    }
-    return n, nil
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	n, ok := d.nodes[id]
+	if !ok {
+		return nil, &ErrNodeNotFound{
+			id,
+		}
+	}
+	return n, nil
 }
 
 func (d *directedGraph[NodeType]) ListNodes() map[string]Node[NodeType] {
-    d.lock.Lock()
-    defer d.lock.Unlock()
-
-    result := map[string]Node[NodeType]{}
-    for nodeID, n := range d.nodes {
-        result[nodeID] = n
-    }
-    return result
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	result := map[string]Node[NodeType]{}
+	for nodeID, n := range d.nodes {
+		result[nodeID] = n
+	}
+	return result
 }
 
 func (d *directedGraph[NodeType]) ListNodesWithoutInboundConnections() map[string]Node[NodeType] {
-    d.lock.Lock()
-    defer d.lock.Unlock()
-
-    result := map[string]Node[NodeType]{}
-    for nodeID, n := range d.nodes {
-        if len(d.connectionsToNode[nodeID]) == 0 {
-            result[nodeID] = n
-        }
-    }
-    return result
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	result := map[string]Node[NodeType]{}
+	for nodeID, n := range d.nodes {
+		connections := d.connectionsToNode[nodeID]
+		if len(connections) == 0 {
+			result[nodeID] = n
+		}
+	}
+	return result
+}
+
+// Validates the specified node IDs and confirms that a connection between them
+// would be valid, then sets the `to` and `from` connections and adds the
+// dependency to the `to` node.
+func (d *directedGraph[NodeType]) connectNodes(fromID, toID string, dependencyType DependencyType) error {
```

Comment:
It would be excellent to record [this bit of wisdom](https://github.com/arcalot/go-dgraph/pull/15#discussion_r1628503552) here or elsewhere (possibly at `node#ConnectDependency()` and also in its interface declaration):
> A node is resolved and becomes ready to process when all of its AND dependencies and at least one of its OR dependencies are resolved.

While this might seem obvious, when there is only one OR dependency or only one AND dependency, the situation is unclear without this statement.

(Perhaps it should be homed where we declare `DependencyType`, but I think this spot is probably better.)